source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#if defined(DEAL_II_WITH_ADOLC) || defined(DEAL_II_TRILINOS_WITH_SACADO)
+
+#  include <deal.II/differentiation/ad/ad_drivers.h>
+#  include <deal.II/differentiation/ad/ad_helpers.h>
+
+#  include <type_traits>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Differentiation
+{
+  namespace AD
+  {
+    /* -------------------------- ADHelperBase -------------------------- */
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    std::set<types::tape_index>
+      ADHelperBase<dim, ADNumberTypeCode, ScalarType>::registered_tapes;
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    dealii::types::global_dof_index
+      ADHelperBase<dim, ADNumberTypeCode, ScalarType>::n_helpers = 0;
```

Comment:
this is the place where you use the type mentioned in the previous comment above in the definition of the variable